File: Crio/crio_programming_interview_problems-master/DecodeWays/DecodeWays.py
def numDecodings(digits: str) -> int: 
# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
	n= len(digits)
	mod = 1000000007
	if digits[0] == '0':
		return 0
	count = [0] * (n + 1); 
	count[0] = 1; 
	count[1] = 1; 
	for i in range(2, n + 1): 
		count[i] = 0; 
		if (digits[i - 1] > '0'): 
			count[i] = count[i - 1]%mod; 
		if (digits[i - 2] == '1' or
		(digits[i - 2] == '2' and
			digits[i - 1] < '7') ): 
			count[i] =((count[i]%mod)+(count[i - 2]%mod))%mod; 
	return count[n]; 
# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

if __name__ == '__main__':
    s = input()
    result = numDecodings(s)
    print(result)
    

File: Crio/crio_programming_interview_problems-master/DecodeWays/DecodeWays.java
import java.io.*;
import java.util.*;
class DecodeWays {
    public int numDecodings(String s) { 
        // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
        int n = s.length();        
        int ways[] = new int[n];
        for(int i =0 ; i < n ; i++)
            ways[i] = -1;
        int answer  = findWays(0, n , ways , s);
        return answer; 
        // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
    }
    // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    public int findWays(int index , int n ,  int ways[] , String s) {
        int mod = 1000000007;
        if(index <= n-1 && s.charAt(index) == '0') {
            return 0;
        }      
        if(index >= n-1) {
            return 1;
        }
        if(ways[index] != -1)
            return ways[index];
        int ans = findWays(index + 1 , n , ways , s)%mod;
        int firstDigit = s.charAt(index) - '0';
        int secondDigit = s.charAt(index + 1) - '0';
        if(10*firstDigit + secondDigit <= 26)
            ans = (ans%mod + findWays(index + 2 , n , ways , s)%mod)%mod;
        ways[index] = ans%mod;
        return ways[index];
    }
    // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
    public static void main(String args[]) {
        Scanner scanner = new Scanner(System.in);
        String s = scanner.nextLine();
        scanner.close();
        int result = new DecodeWays().numDecodings(s);
        System.out.println(result);
     }
}
